[V2V] Filter out jobs.migration_task.conversion_host.nil? in InfraConversionThrottler.running_conversions

fdupont-redhat opened the pull request:
From field feedback, it seems that it can happen that a migration task which is running doesn't have a conversion host (mostly because it has been deleted). The consequence is that the InfraConversionThrottler cannot enforce the limits, as an exception is raised when trying to connect to a nullified conversion host. The symptoms are this kind of log message:

```
[----] I, [2020-02-26T03:06:07.440488 #4518:2b12ad6985bc]  INFO -- : Q-task_id([job_dispatcher]) Exception in realtime_block :total_time - Timings: {:v2v_dispatching=>0.0015590190887451172, :v2v_limits=>0.019486188888549805, :total_time=>0.02106475830078125}
[----] E, [2020-02-26T03:06:07.444027 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) MIQ(MiqQueue#deliver) Message id: [1000000918612], Error: [undefined method `cpu_limit' for nil:NilClass]
[----] E, [2020-02-26T03:06:07.444194 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) [NoMethodError]: undefined method `cpu_limit' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
[----] E, [2020-02-26T03:06:07.444247 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) /var/www/miq/vmdb/lib/infra_conversion_throttler.rb:83:in `block in apply_limits'
/var/www/miq/vmdb/lib/infra_conversion_throttler.rb:80:in `each'
/var/www/miq/vmdb/lib/infra_conversion_throttler.rb:80:in `apply_limits'
/var/www/miq/vmdb/app/models/job_proxy_dispatcher.rb:108:in `apply_v2v_limits'
```

This pull request rejects the running jobs that have a migration task without a conversion host from the list of running jobs. It also logs an error message.

fdupont-redhat commented:
@miq-bot add-label v2v, bug, ivanchuk/yes
@miq-bot add-reviewer @agrare

agrare commented:
@fdupont-redhat can you add a spec test covering this case?

agrare commented on lib/infra_conversion_throttler.rb:
```diff
@@ -62,7 +62,7 @@ def self.pending_conversion_jobs
   # @return [Hash] the list of jobs in state 'running', grouped by conversion host
   def self.running_conversion_jobs
-    running = InfraConversionJob.where.not(:state => ['waiting_to_start', 'finished'])
+    running = InfraConversionJob.where.not(:state => ['waiting_to_start', 'finished']).reject { |job| job.migration_task&.conversion_host.nil? }
```
@fdupont-redhat since this is a case that shouldn't happen I wonder if it'd be better to log an error [here](https://github.com/ManageIQ/manageiq/blob/9085751ba7bae6fb5ad86080719ba5b57017840c/lib/infra_conversion_throttler.rb#L85) with the jobs that don't have a conversion host.

agrare commented:
This was the error BTW:
```
[----] I, [2020-02-26T03:06:07.440488 #4518:2b12ad6985bc]  INFO -- : Q-task_id([job_dispatcher]) Exception in realtime_block :total_time - Timings: {:v2v_dispatching=>0.0015590190887451172, :v2v_limits=>0.019486188888549805, :total_time=>0.02106475830078125}
[----] E, [2020-02-26T03:06:07.444027 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) MIQ(MiqQueue#deliver) Message id: [1000000918612], Error: [undefined method `cpu_limit' for nil:NilClass]
[----] E, [2020-02-26T03:06:07.444194 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) [NoMethodError]: undefined method `cpu_limit' for nil:NilClass  Method:[block (2 levels) in <class:LogProxy>]
[----] E, [2020-02-26T03:06:07.444247 #4518:2b12ad6985bc] ERROR -- : Q-task_id([job_dispatcher]) /var/www/miq/vmdb/lib/infra_conversion_throttler.rb:83:in `block in apply_limits'
/var/www/miq/vmdb/lib/infra_conversion_throttler.rb:80:in `each'
/var/www/miq/vmdb/lib/infra_conversion_throttler.rb:80:in `apply_limits'
/var/www/miq/vmdb/app/models/job_proxy_dispatcher.rb:108:in `apply_v2v_limits'
```

fdupont-redhat commented:
@agrare so, the error is different from what I was checking. I've updated the PR with a check on the conversion host existence, and it also logs. I've also updated the specs accordingly.

agrare commented:
Looks good, you just need to fix the spec failures

agrare commented on lib/infra_conversion_throttler.rb:
```diff
@@ -80,6 +80,11 @@ def self.running_conversion_jobs
   def self.apply_limits
     running_conversion_jobs.each do |ch, jobs|
+      if ch.nil?
+        _log.error("Migration task for '#{migration_task.source.name}' is running but has no conversion host")
```
`migration_task.source.name` <- is the issue here, there's no `migration_task` yet here.

If you want to list the names of all of them you could do something like `source_names = jobs.map { |job| job.migration_task&.source&.name }.compact.join(", ")` and if you expect this could be a lot of vms you can limit it with `.compact.take(max_number_to_print)`

fdupont-redhat commented:
@miq-bot add-label jansa/yes?

simaishi commented:
Jansa backport details:

```
$ git log -1
commit e1ff082940e1602252024d5534818cacee1744a0
Author: Adam Grare <agrare@redhat.com>
Date:   Fri Feb 28 15:37:55 2020 -0500

    Merge pull request #19890 from fdupont-redhat/v2v_check_chost_in_throttler_apply_limits

    [V2V] Filter out jobs.migration_task.conversion_host.nil? in InfraConversionThrottler.running_conversions

    (cherry picked from commit da451b6898dfa29673d8b216239d43672bcae77c)

    https://bugzilla.redhat.com/show_bug.cgi?id=1819998
```

simaishi commented:
@fdupont-redhat Backporting to ivanchuk conflicts. Please create a separate PR. I realized this PR should have been backported before https://github.com/ManageIQ/manageiq/pull/20105. Not sure if that's the reason Travis is red right now. Sorry if that's the case...

fdupont-redhat commented:
@simaishi here is the backport PR: https://github.com/ManageIQ/manageiq/pull/20182

simaishi commented:
Backported to ivanchuk via https://github.com/ManageIQ/manageiq/pull/20182
